### PR TITLE
Update etcdadm config type to have separate configuration for formats

### PR DESCRIPTION
**Pull request by g-gaston**
* Separate structs for bottlerocket and cloud-config
* Remove hardcoded image for BR bootstrap and pass it as configuration
This is not necessarily the best api design, but trying to minimize changes here
Some of the code in the controller and bootstrap user-data generators can probably use a bit of a refactor as well

### CODE REVIEW

File: api/v1alpha3/etcdadmconfig_types.go [hunk 1]
```diff
@@ -44,6 +44,30 @@ type EtcdadmConfigSpec struct {
...
+	// +optional
+	Format Format `json:"format,omitempty"`
+
+	// BottlerocketConfig specifies the configuration for the bottlerocket bootstrap data
+	// +optional
+	BottlerocketConfig *BottlerocketConfig `json:"bottlerocketConfig,omitempty"`
+
+	// CloudConfigConfig specifies the configuration for the cloud-init bootstrap data
+	// +optional
+	CloudConfigConfig *CloudConfigConfig `json:"cloudConfigConfig,omitempty"`
```
Comment (mrajashree) [inline api/v1alpha3/etcdadmconfig_types.go]: can we rename this to CloudInitConfig?

File: api/v1alpha3/etcdadmconfig_types.go [hunk 2]
```diff
@@ -44,6 +44,30 @@ type EtcdadmConfigSpec struct {
...
+}
+
+type BottlerocketConfig struct {
+	// EtcdImage specifies the etcd image to use by etcdadm
+	EtcdImage string `json:"etcdImage,omitempty"`
+
+	// BootstrapImage specifies the container image to use for bottlerocket's bootstrapping
+	BootstrapImage string `json:"bootstrapImage"`
+}
+type CloudConfigConfig struct {
```
Comment (mrajashree) [inline api/v1alpha3/etcdadmconfig_types.go]: For CAPV we run commands to set hostname through PreEtcdadmCommands, this is done for CAPV through the KCP spec too. Do we not need those commands for BR bootstrap? 
Even if we don't, I think we should keep Pre and PostEtcdadmCommands as top level fields along with Users on the EtcdadmConfigSpec

File: controllers/etcdadmconfig_controller.go [hunk 3]
```diff
@@ -292,7 +288,7 @@ func (r *EtcdadmConfigReconciler) joinEtcd(ctx context.Context, scope *Scope) (_
 	log.Info("Machine Controller has set address on init machine")
 
 	etcdCerts := etcdCACertKeyPair()
-	rerr = etcdCerts.Lookup(
+	_ = etcdCerts.Lookup(
```
Comment (mrajashree) [inline controllers/etcdadmconfig_controller.go]: I realize this code was swallowing the error earlier too, but if we encounter error here can we return it?

File: pkg/userdata/bottlerocket/prepare.go [hunk 4]
```diff
@@ -28,6 +30,22 @@ func patchCertPaths(input *userdata.BaseUserData) {
...
+	repository, tag := splitRepositoryAndTag(config.EtcdImage)
+	return userdata.EtcdadmArgs{
+		Version:         strings.TrimPrefix(tag, "v"), // trim "v" to get pure simver because that's what etcdadm expects.
+		ImageRepository: repository,
+	}
+}
+
+func splitRepositoryAndTag(image string) (repository, tag string) {
+	lastInd := strings.LastIndex(image, ":")
+	if lastInd == -1 || lastInd == len(image)-1 {
```
Comment (mrajashree) [inline pkg/userdata/bottlerocket/prepare.go]: the first part of if condition where `:` is not a part of image at all will return the image repo correctly, but the second part that checks if the `:` was at the end will return image containing the `:` at the end, so I think it should be:
```suggestion
	if lastInd == -1 {
		return image, ""
	}
	if lastInd == len(image)-1 {
		return image[:lastInd], ""
	}
```

File: api/v1alpha3/etcdadmconfig_types.go (hunk 2 above)
Comment (g-gaston) [reply api/v1alpha3/etcdadmconfig_types.go]: Done

File: api/v1alpha3/etcdadmconfig_types.go (hunk 1 above)
Comment (g-gaston) [reply api/v1alpha3/etcdadmconfig_types.go]: done

File: controllers/etcdadmconfig_controller.go (hunk 3 above)
Comment (g-gaston) [reply controllers/etcdadmconfig_controller.go]: done!

**Pull request merged**